plura-ai/plura
# feat(api): add rate-limiting with Upstash for dev and prod environments

## Description

Implemented rate-limiting per IP using Upstash Redis. This ensures that API endpoints are protected against excessive requests. The rate limit feature is operational in both development and production environments.

## Implementation Details

- Used `@upstash/ratelimit` library to implement rate-limiting
- Excluded the following routes from rate-limiting:
  `/v1/status`
  `/v1/health`
- All other routes are processed through the rate-limiting handler.
- The current configuration allows 100 requests in development and 60 requests in production per minute per IP.

## Testing

- Endpoint: `http://localhost:3001/v1/session`
- Steps:
  - Set the rate limit to 5 requests in development per minute for testing purposes.
  - Make more than 5 requests to the endpoint within a minute.
  - Observe the response: A `429` status code with a message `You hit the rate limit`

## Screenshots / Demo

![image](https://github.com/user-attachments/assets/70d9fed2-6b10-41c1-8ccd-3ae470334151)

## Checklist

Please confirm that you have completed the following:



- [x] I have provided a clear and concise description of the feature implementation.
- [x] I have followed the code style and formatting guidelines.
- [x] I have added any necessary documentation or comments.
- [ ] I have updated any relevant tests or test cases.

#### Check only those which are applicable.

**Documentation Updated?**

- [ ] If your feature adds a new API, then documentation related to the feature like API-Contracts, and Data-models must be updated.

**Breaking Changes**

- [ ] If your feature introduces breaking changes or if something is missing, please mention the related issue tickets.

**Development Tested?**

- [x] Confirm whether the changes have been tested locally during development.

**Tested in Staging?**

- [ ] Indicate whether the changes have been tested in the staging environment.

**Under Feature Flag**

- [ ] Specify if the changes are currently under a feature flag.

**Database Changes**

- [ ] Indicate whether the changes include modifications to the database.

## Issue ID

#102
